File: Xit/XTPreviewController.m
#import "XTPreviewController.h"
#import "XTConstants.h"
#import "XTPreviewItem.h"
#import "XTRepository.h"
#import <Quartz/Quartz.h>

@interface XTPreviewController ()

@property QLPreviewView *view;

@end


@implementation XTPreviewController

@dynamic view;

- (void)clear
{
  self.view.previewItem = nil;
}

- (void)loadPath:(NSString *)path
           model:(id<XTFileChangesModel>)model
          staged:(BOOL)staged
{
  if (!staged) {
    self.view.previewItem = [model unstagedFileURL:path];
  }
  else {
    XTPreviewItem *previewItem = (XTPreviewItem*)self.view.previewItem;
    
    if (![previewItem isKindOfClass:[XTPreviewItem class]]) {
      previewItem = [[XTPreviewItem alloc] init];
      self.view.previewItem = previewItem;
    }
    
    previewItem.model = model;
    previewItem.path = path;
    [self.view refreshPreviewItem];
  }
}

- (BOOL) canSetWhitespace
{
  return NO;
}

- (BOOL) canSetTabWidth
{
  return NO;
}

@end

File: Xit/XTPreviewController.swift
import Foundation
import Quartz

/// Controller for the QuickLook preview tab.
class XTPreviewController: NSViewController
{
  var isLoaded: Bool = false
  
  var qlView: QLPreviewView { return view as! QLPreviewView }
  
  override func awakeFromNib()
  {
    // Having the QL view in the xib was causing odd problems
    view = QLPreviewView(frame: NSRect(x: 0, y: 0, width: 50, height: 50),
                         style: .normal)
  }
  
  func refreshPreviewItem()
  {
    qlView.refreshPreviewItem()
  }
}

extension XTPreviewController: XTFileContentController
{
  public func clear()
  {
    qlView.previewItem = nil
    isLoaded = false
  }
  
  public func load(path: String!, fileList: FileListModel)
  {
    let previewView = qlView
  
    if fileList is WorkspaceFileList {
      guard let urlString = fileList.fileURL(path)?.absoluteString
      else {
        previewView.previewItem = nil
        isLoaded = true
        return
      }
      // Swift's URL doesn't conform to QLPreviewItem because it's not a class
      let nsurl = NSURL(string: urlString)
    
      DispatchQueue.main.async {
        previewView.previewItem = nsurl
        self.isLoaded = true
      }
    }
    else {
      var previewItem: PreviewItem! = previewView.previewItem
        as? PreviewItem
      
      DispatchQueue.main.async {
        if previewItem == nil {
          previewItem = PreviewItem()
          previewView.previewItem = previewItem
        }
        previewItem.fileList = fileList
        previewItem.path = path
        previewView.refreshPreviewItem()
        self.isLoaded = true
      }
    }
  }
}
